apis/v1beta1/httproute_types.go

```diff
@@ -156,7 +156,9 @@ type HTTPRouteRule struct {
 	// ties. Across all rules specified on applicable Routes, precedence must be
 	// given to the match with the largest number of:
 	//
-	// * Characters in a matching path.
+	// * Characters in a matching "Exact" path match
+	// * Characters in a matching "RegularExpression" path match
```

Comment:
Regex between Exact and Prefix feels a bit odd...

TBH, using Characters at all in regex feels odd entirely. Characters has no meaning in regex. is there any proxy implementation that does this today? I worry we will see users inserting dummy characters

Reply:
contour uses regex between exact and prefix because a regex "could be" more specific than a prefix match, not a perfect heuristic but yeah

also yeah we do sort our regex matches on length just to enable consistent ordering, again not a perfect heuristic

Reply:
Chatting with @briantkennedy about this and he pointed out that this would likely not be compatible with NGINX where RegEx matches always happen last. It's probably best to require "Exact", "Prefix", then "RegEx". @brianehlert @pleshakov are probably best to confirm from NGINX side though.

Reply:
I am not quite sure for putting the order of regex to where, since it is different in implementations. I just added conformance to test:
```
Characters in a matching "Exact" path match
Characters in a matching "Prefix" path match
```

Reply:
I don't think we're going to be able to mandate any specific precedence for RegEx path matches in the API. Instead, we probably need to label that as implementation-specific. Looking at the [NGINX reference docs], it does not look like it would be possible for it to be conformant with any precedence based on the length of a regex.

>To find location matching a given request, nginx first checks locations defined using the prefix strings (prefix locations). Among them, the location with the longest matching prefix is selected and remembered. Then regular expressions are checked, in the order of their appearance in the configuration file. The search of regular expressions terminates on the first match, and the corresponding configuration is used. If no match with a regular expression is found then the configuration of the prefix location remembered earlier is used.

Since we already consider RegEx path matching "Implementation-Specific" we likely should also consider the precedence of RegEx path matches "Implementation-Specific".

I'd recommend removing the "RegEx" mention from the list and adding a note below the list that states that "The precedence of RegularExpression path matches are implementation-specific."

Reply:
Thanks @robscott